lib/Dialect/Pipeline/PipelineOps.cpp

```diff
@@ -327,6 +327,154 @@ LogicalResult ReturnOp::verify() {
+// regs($register : type($register) (`gated by` `[` $clockGates `]`)?, ...)
+ParseResult parseStageRegisters(
+    OpAsmParser &parser,
+    llvm::SmallVector<OpAsmParser::UnresolvedOperand, 4> &registers,
+    llvm::SmallVector<mlir::Type, 1> &registerTypes,
+    llvm::SmallVector<OpAsmParser::UnresolvedOperand, 4> &clockGates,
+    ArrayAttr &clockGatesPerRegister) {
+
+  if (failed(parser.parseOptionalKeyword("regs"))) {
+    clockGatesPerRegister = parser.getBuilder().getI64ArrayAttr({});
+    return success(); // no registers to parse.
+  }
+
+  llvm::SmallVector<int64_t> clockGatesPerRegisterList;
+  if (failed(parser.parseCommaSeparatedList(AsmParser::Delimiter::Paren, [&]() {
+        OpAsmParser::UnresolvedOperand v;
+        Type t;
+        llvm::SmallVector<OpAsmParser::UnresolvedOperand> cgs;
+        if (parseSingleStageRegister(parser, v, t, cgs))
+          return failure();
+        registers.push_back(v);
+        registerTypes.push_back(t);
+        llvm::append_range(clockGates, cgs);
+        clockGatesPerRegisterList.push_back(cgs.size());
+        return success();
+      })))
+    return failure();
+
+  clockGatesPerRegister =
+      parser.getBuilder().getI64ArrayAttr(clockGatesPerRegisterList);
+
+  return success();
+}
+
+void printStageRegisters(OpAsmPrinter &p, Operation *op, ValueRange registers,
+                         TypeRange registerTypes, ValueRange clockGates,
+                         ArrayAttr clockGatesPerRegister) {
+  if (registers.empty())
+    return;
+
+  p << "regs(";
+  size_t clockGateStartIdx = 0;
+  llvm::interleaveComma(
+      llvm::zip(registers, registerTypes, clockGatesPerRegister), p,
+      [&](auto it) {
+        auto &[reg, type, nClockGatesAttr] = it;
+        p << reg << " : " << type;
+        int64_t nClockGates =
+            nClockGatesAttr.template cast<IntegerAttr>().getInt();
+        if (nClockGates == 0)
+          return;
+        p << " gated by [";
+        llvm::interleaveComma(clockGates.slice(clockGateStartIdx, nClockGates),
+                              p);
+        p << "]";
+        clockGateStartIdx += nClockGates;
+      });
+  p << ")";
+}
+
+void StageOp::build(OpBuilder &odsBuilder, OperationState &odsState,
+                    Block *dest, ValueRange registers,
+                    ValueRange passthroughs) {
+  odsState.addSuccessors(dest);
+  odsState.addOperands(registers);
+  odsState.addOperands(passthroughs);
+  odsState.addAttribute("operand_segment_sizes",
+                        odsBuilder.getDenseI32ArrayAttr(
+                            {static_cast<int32_t>(registers.size()),
+                             static_cast<int32_t>(passthroughs.size()),
+                             /*clock gates*/ static_cast<int32_t>(0)}));
+  llvm::SmallVector<int64_t> clockGatesPerRegister(registers.size(), 0);
+  odsState.addAttribute("clockGatesPerRegister",
+                        odsBuilder.getI64ArrayAttr(clockGatesPerRegister));
+}
+
+void StageOp::build(
```

Comment:
This builder isn't used anywhere, correct? Why add it pre-emptively? I don't like untested code.